V8 - LogViewer First Draft Doc (#1524)
umbraco/UmbracoDocs: Getting-Started/LogViewer/index-v8.md

---
keywords: logging logviewer logs serilog messagetemplates logs v8 version8
versionFrom: 8.0.0
---

# Log Viewer

> From version 8 we are now shipping a built in Log Viewer feature, to allow you to filter and view log entries and perform much more complex search queries to help you find the log entries that you are interested in your Umbraco site.

Reviewer: ```suggestion
From version 8 we are now shipping a built in Log Viewer feature, to allow you to filter and view log entries and perform much more complex search queries to help you find the log entries that you are interested in your Umbraco site.
You can find the log viewer in the settings section.
```

## Benefits

> Wanted to find all log entries which contains the same reqest ID or find all items in the log where a property called duration is greater than 1000ms. With the power of structured logging and a query language we are able to search and find log items for very specific scenarios and when debugging the client site you should now have more power to see and find patterns in your log files and get rid of those pesky errors.

Reviewer: ```suggestion
Have you ever wanted to find all log entries which contains the same request ID or find all items in the log where a property called duration is greater than 1000ms? 
With the power of structured logging and a query language we are able to search and find log items for very specific scenarios. When debugging the client site you should now have more power to see and find patterns in your log files and get rid of those pesky errors.
```

## Example queries
Here are a handful example queries to get you started, however the saved searches contain some further examples. For more details on the syntax head over to the https://github.com/serilog/serilog-filters-expressions project.

**Find all logs that are from the namespace 'Umbraco.Core'**<br/>
`StartsWith(SourceContext, 'Umbraco.Core')`<br/>

> **Find all logs that have the property 'Duration' and the duration is greater than 1000ms'**<br/>

Reviewer: ```suggestion
**Find all logs that have the property 'Duration' and the duration is greater than 1000ms**<br/>
```

`Has(Duration) and Duration > 1000`<br/>

> **Find all logs that the message has localhost in it with SQL like**<br/>

Reviewer: ```suggestion
**Find all logs where the message has localhost in it with SQL like**<br/>
```

`@Message like '%localhost%'`<br/>

## Saved Searches
When writing a custom query that you wish to use often, it is possible to save this and use the dropdown to re-use your saved search. To add a new saved search, use the search box to type your query and click the star icon. In doing so you can give it a friendly name. The default location of the saved searches are persisted as JSON to the following file: `/config/logviewer.searches.config.js`

## Implementing your own Log Viewer

> With the flexability of Umbraco, we give you the power to implement your own `ILogViewer` where you are able to fetch logs and the saved searched from a different location such as Azure table storage.

Reviewer: ```suggestion
With the flexibility of Umbraco, we give you the power to implement your own `ILogViewer` where you are able to fetch logs and the saved searched from a different location such as Azure table storage.
```

### Create your own implementation

> To easily do this we can implment a base class `LogVieweSourceBase` from `Umbraco.Core.Logging.Viewer` like so.

Reviewer: ```suggestion
To easily do this we can implement a base class `LogViewerSourceBase` from `Umbraco.Core.Logging.Viewer` like so.
```

namespace My.Website
{
    [RuntimeLevel(MinLevel = RuntimeLevel.Run)]
    public class LogViewerSavedSearches : IUserComposer
    {
        public void Compose(Composition composition)
        {
            composition.SetLogViewer<AzureTableLogViewer>();
        }
    }
}
```

### Configure Umbraco to log to Azure Table Storage
Now with the above two classes we have the plumbing in place to view logs from an Azure Table, however we are not persisting our logs into the Azure table storage account. So we need to configure the Serilog logging pipeline to store our logs into Azure table storage.

* Install Serilog.Sinks.AzureTableStorage from Nuget (Latest pre-release)
* Update `config/serilog.user.config` with credentials (so logs persist to Azure)

```xml
<add key="serilog:using:AzureTableStorage" value="Serilog.Sinks.AzureTableStorage" />
<add key="serilog:write-to:AzureTableStorage.connectionString" value="DefaultEndpointsProtocol=https;AccountName=ACCOUNT_NAME;AccountKey=KEY;EndpointSuffix=core.windows.net" />
<add key="serilog:write-to:AzureTableStorage.formatter" value="Serilog.Formatting.Compact.CompactJsonFormatter, Serilog.Formatting.Compact" />
```

> For more in depth information about logging and how to configure it, please read the logging documentation

Reviewer: ```suggestion
For more in depth information about logging and how to configure it, please read the [logging documentation](../Code/Debugging/Logging/index-v8).
```

Author: Not sure why I can't use the Commit suggestion feature here - will manually make the change @jmayntzhusen
